src/modules/land_detector/MulticopterLandDetector.cpp

```diff
@@ -181,12 +183,29 @@ bool MulticopterLandDetector::_get_ground_contact_state()
 
 	// if we have a valid velocity setpoint and the vehicle is demanded to go down but no vertical movement present,
 	// we then can assume that the vehicle hit ground
-	_in_descend = _is_climb_rate_enabled()
-		      && (_vehicle_local_position_setpoint.vz >= land_speed_threshold);
+	const bool vz_valid = (lpos_available && _vehicle_local_position.v_z_valid);
+
+	if (_vehicle_control_mode.flag_control_climb_rate_enabled && vz_valid) {
+		// setpoints can briefly be NAN to signal resets, TODO: fix in multicopter position controller
+		if (PX4_ISFINITE(_vehicle_local_position_setpoint.vz)) {
```

Comment:
This means in every case where there's no vertical velocity setpoint the `_in_descend` state just stays how it was. The only cases that come to my mind where no vertical velocity setpoint is set (intentionally) are: the thrust cut when ground is detected (that's the constelation that leads to the bug you found) and the in failsafe descend when there's no RC and the vehicle looses the vertical velocity estimate.

What about we capture the case where there's no vertical velocity setpoint and in that case compare vertical acceleration. E.g. the thrust cut is a huge downwards acceleration to enable the land detector to detect maybe landed and landed. And with that huge downwards acceleration you expect no freefall.

Reply:
The acceleration condition sounds good, I'm not sure what that wasn't already a factor. My only immediate concern is the current time scale of the ground contact and its hysteresis.

Reply:
In `flag_control_climb_rate_enabled` (and vz_valid) if the vz setpoint is NAN we could just set _in_descend to true.

Reply:
> in failsafe descend when there's no RC and the vehicle looses the vertical velocity estimate.

Actually, this one isn't going to matter because `ground_contact` is only used by `maybe_landed` if vz and z estimates are valid.